lib/active_record/connection_adapters/sqlserver/database_statements.rb

```diff
@@ -140,6 +140,37 @@ def default_insert_value(column)
         private :default_insert_value
 
         def build_insert_sql(insert) # :nodoc:
+          if insert.skip_duplicates?
+            # Do we have a unique_by index? Use index columns
+            conflict_columns = if (unique_by = insert.send(:insert_all).unique_by)
```

Comment:
Can we not do `insert.insert_all`?

Reply:
Suggest we cache insert_all in a local variable rather than have to look it up each time.

Reply:
insert_all is private, it's not possible to do `insert.insert_all`. 
Thanks for the suggestion but I'm still not paying attention to things like this. As I said, the PR is in an early stage. I have plans to improve the code (refactor) but right now I'm focused on produce the SQL that cover all tests.